fix: Update time picker dialog input size

rkishan516 opened the pull request:
fix: Update time picker dialog input size
Fixes: #162796 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

rkishan516 committed:
fix: Update time picker dialog input size

MitchellGoodwin reviewed (approved):
LGTM! Thank you for making this PR. My main concern was accessibility, and the text being scaled up. But there are clamping on the text scaling and specifically in the input mode we have room to increase the height a little, so this should be fine. Due to box constraints, this only affects the max size and will not affect input dialogs that do not need the extra space.

justinmc reviewed (approved):
LGTM assuming Google tests pass 👍 . I didn't realized this height would be hardcoded but I guess so.

MitchellGoodwin commented:
> LGTM assuming Google tests pass 👍 . I didn't realized this height would be hardcoded but I guess so.

The material date pickers have hard coded maxes due to containing text scaling. The calendar view in particular has to watch out for a lot of text being scaled up and filling the screen. There's wiggle room with the input picker though, which has less text.

MitchellGoodwin commented:
I was wrong, Google testing is showing that the pickers are expanding when they don't need to. It doesn't look great. I'm not sure what the best solution is.

rkishan516 commented:
> I was wrong, Google testing is showing that the pickers are expanding when they don't need to. It doesn't look great. I'm not sure what the best solution is.

I will see, what other solution I can come up with.

rkishan516 commented:
@MitchellGoodwin Few solutions
We can make actions horizontally or vertically scrollable.
or
We can change dialog size dynamically with content size.

MitchellGoodwin commented:
> @MitchellGoodwin Few solutions We can make actions horizontally or vertically scrollable. or We can change dialog size dynamically with content size. (As you said text scaling will be an issue).

Making them scrollable is nice because it will avoid errors, but would lead to bad UI, in my opinion. We could do that as a last resort.

Changing the dialog size dynamically would be preferable. For the input time picker there's a lot of room to grow. All the careful constraints are target towards the calendar date picker, which fills up much more space. I'd have to double check but I believe font size can only increase by 10%. When I looked at the code, dynamically increasing the size does seem difficult, however, with the expanded widgets involved.

rkishan516 commented:
> > @MitchellGoodwin Few solutions We can make actions horizontally or vertically scrollable. or We can change dialog size dynamically with content size. (As you said text scaling will be an issue).
> 
> Making them scrollable is nice because it will avoid errors, but would lead to bad UI, in my opinion. We could do that as a last resort.
> 
> Changing the dialog size dynamically would be preferable. For the input time picker there's a lot of room to grow. All the careful constraints are target towards the calendar date picker, which fills up much more space. I'd have to double check but I believe font size can only increase by 10%. When I looked at the code, dynamically increasing the size does seem difficult, however, with the expanded widgets involved.

Yaa, I have converted `Expanded` into `Flexible` and given a top padding for actions same as dialog's bottom padding, such that it looks symmetrical. And since dial mode needs some height, so I have added maxHeight constraints. So, its dynamic till a point.

rkishan516 commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -2507,7 +2507,10 @@ class _TimePickerDialogState extends State<TimePickerDialog> with RestorationMix
-      padding: EdgeInsetsDirectional.only(start: theme.useMaterial3 ? 0 : 4),
+      padding: EdgeInsetsDirectional.only(
+        start: theme.useMaterial3 ? 0 : 4,
+        top: (pickerTheme.padding ?? defaultTheme.padding).vertical / 2,
```
I think adding `Flexible` for spacing will start taking entire allowed height and which is not what we want. I wanted spacing around actions symmertical. So, I will change documentation in picker theme.

MitchellGoodwin replied:
I think a flexible spacing widget could be added if a sub `Column` was added as well, setting the priority of how widgets grow.

Some of the checks are failing for overflow errors, so I imagine that since this space is static and cannot shrink it is lowering the amount of open space for the rest of the widgets.

MitchellGoodwin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -2606,13 +2606,15 @@ class _TimePickerDialogState extends State<TimePickerDialog> with RestorationMix
-                  height: allowedSize.height,
                   duration: _kDialogSizeAnimationDuration,
                   curve: Curves.easeIn,
+                  constraints: BoxConstraints(maxHeight: allowedSize.height),
```
Apologies for letting this sit, I needed to find time to look into it. It still looks like it's expanding a bit too much when it doesn't need to.

What if we added a minHeight to this with the original minimum height `216`, then added a mainAxisAlignment of `spaceBetween` to the Column and removed the spacer and flex from the Flexible.

We want to: have a small space between the form and actions if possible. If the form expands, then we want the space to go away first until it's gone completely, _then_ start growing the dialog box.

Theoretically this combo would work. What do you think?

rkishan516 replied:
Yup, I think this should work. Since we are doing spaceBetween so first that will shrink the space and then it will change dialog height.

rkishan516 replied:
@MitchellGoodwin I have pushed the change, Just one change from your suggestion i.e. keeping it limited to input and inputOnly mode.

MitchellGoodwin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -2606,33 +2606,45 @@ class _TimePickerDialogState extends State<TimePickerDialog> with RestorationMix
-                            switchToTimerEntryModeIcon: widget.switchToTimerEntryModeIcon,
-                          ),
-                        ),
+                      Builder(
```
What's the benefit of using a Builder? To be able to do the _entryMode check below in less lines?

rkishan516 replied:
Yes, @MitchellGoodwin.

MitchellGoodwin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -2606,33 +2606,45 @@ class _TimePickerDialogState extends State<TimePickerDialog> with RestorationMix
                   duration: _kDialogSizeAnimationDuration,
                   curve: Curves.easeIn,
+                  constraints: BoxConstraints(
+                    minHeight: _kTimePickerMinInputSize.height,
```
This should be the original value. So above we'd want:
```dart
static const Size _kTimePickerInputSize = Size(312, 252);
static const double _kTimePickerInputMinimumHeight = 216;
```

then here it would be:
```dart
constraints: BoxConstraints(
  minHeight: _kTimePickerInputMinimumHeight,
  maxHeight: allowedSize.height,
),
```
This keeps dialogs that do not need to expand to not be bigger than they were before.

rkishan516 replied:
Sorry, My bad I didn't checked value of _kTimePickerMinInputSize and thought it will be 216. I have pushed required change.

MitchellGoodwin reviewed (approved):
LGTM! Thank you for the follow up fixes.
